# Author's Details:-
Ayan Chowdhury  
ayanchowdhury7@gmail.com  
8900632230


# Project

Machine Learning model , which will help the management to shortlist the potential candidates while applying. The Machine Learning model will be built on a given data-set.


In [1]:
#importing modules

import os
import re
import pandas as pd
import numpy as np
import xml

In [2]:
# Loading the data
path = "/home/ice_bear/Downloads/data_science_sample_data_v2.xlsx"
df = pd.read_excel(path)
df

Application_ID Current City  Python (out of 3)  R Programming (out of 3)  \
0           DS0001    Bangalore                  1                         0   
1           DS0002       Mumbai                  2                         1   
2           DS0003       Mumbai                  2                         0   
3           DS0004      Dhanbad                  2                         0   
4           DS0005    Bangalore                  2                         0   
..             ...          ...                ...                       ...   
606         DS0607         Pune                  2                         2   
607         DS0608    Bangalore                  1                         0   
608         DS0609        Delhi                  2                         2   
609         DS0610    Bangalore                  3                         2   
610         DS0611     Thrissur                  2                         2   

     Data Science (out of 3)  \
0                          3   
1                          2   
2                          0   
3                          2   
4                          0   
..                       ...   
606                        2   
607                        0   
608                        0   
609                        2   
610                        0   

                                          Other skills  \
0    Machine Learning, Arduino, C Programming, CSS,...   
1    AutoCAD, MS-Office, Machine Learning, Microsof...   
2    C++ Programming, Data Structures, Image Proces...   
3    Algorithms, C++ Programming, Data Structures, ...   
4                       MS-Word, Python, SQL, MS-Excel   
..                                                 ...   
606  AutoCAD, MS-Excel, Deep Learning, Machine Lear...   
607                        MS-Excel, HTML, Python, SQL   
608  Data Analytics, Machine Learning, Natural Lang...   
609  Data Analytics, Machine Learning, Python, C Pr...   
610  Computer Vision, Data Analytics, Image Process...   

                                             Institute  \
0                         Global Academy of Technology   
1    Aegis School Of Business, Data Science, Cyber ...   
2                                         VJTI, Mumbai   
3                                    IIT (ISM) Dhanbad   
4                                                 Vvce   
..                                                 ...   
606                                   Defour Analytics   
607  International Institute Of Information Technology   
608                                 Quantum University   
609        College Of Applied Science IHRD, Malampuzha   
610                                             Amrita   

                              Degree                                  Stream  \
0      Bachelor of Engineering (B.E)  Electrical and Electronics Engineering   
1                                NaN                                     PGP   
2    Bachelor of Technology (B.Tech)                     Information Systems   
3                  Integrated M.Tech               Mathematics and Computing   
4      Bachelor of Engineering (B.E)           Electronics and Communication   
..                               ...                                     ...   
606    Bachelor of Engineering (B.E)                            Data Science   
607       PG Diploma in Data Science                            Data Science   
608  Bachelor of Technology (B.Tech)                  Mechanical Engineering   
609       Bachelor of Science (B.Sc)        Computer Science And Mathematics   
610    Master of Technology (M.Tech)                 E Learning Technologies   

     Current Year Of Graduation Performance_PG Performance_UG Performance_12  \
0                          2019            NaN        7.73/10            NaN   
1                          2020            NaN      68.00/100            NaN   
2                          2018            NaN        8.85/10  

# Data Cleaning, Data Wrangling and Data Analysis 

In [3]:
df.info

<bound method DataFrame.info of     Application_ID Current City  Python (out of 3)  R Programming (out of 3)  \
0           DS0001    Bangalore                  1                         0   
1           DS0002       Mumbai                  2                         1   
2           DS0003       Mumbai                  2                         0   
3           DS0004      Dhanbad                  2                         0   
4           DS0005    Bangalore                  2                         0   
..             ...          ...                ...                       ...   
606         DS0607         Pune                  2                         2   
607         DS0608    Bangalore                  1                         0   
608         DS0609        Delhi                  2                         2   
609         DS0610    Bangalore                  3                         2   
610         DS0611     Thrissur                  2                         2   

     Da

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 611 entries, 0 to 610
Data columns (total 14 columns):
Application_ID                611 non-null object
Current City                  611 non-null object
Python (out of 3)             611 non-null int64
R Programming (out of 3)      611 non-null int64
Data Science (out of 3)       611 non-null int64
Other skills                  601 non-null object
Institute                     611 non-null object
Degree                        575 non-null object
Stream                        580 non-null object
Current Year Of Graduation    611 non-null int64
Performance_PG                128 non-null object
Performance_UG                533 non-null object
Performance_12                363 non-null object
Performance_10                339 non-null object
dtypes: int64(4), object(10)
memory usage: 67.0+ KB


In [5]:
df.describe()

Python (out of 3)  R Programming (out of 3)  Data Science (out of 3)  \
count         611.000000                611.000000               611.000000   
mean            1.788871                  0.582651                 0.872340   
std             0.780673                  0.918459                 1.014681   
min             0.000000                  0.000000                 0.000000   
25%             1.000000                  0.000000                 0.000000   
50%             2.000000                  0.000000                 0.000000   
75%             2.000000                  1.000000                 2.000000   
max             3.000000                  3.000000                 3.000000   

       Current Year Of Graduation  
count                  611.000000  
mean                  2018.487725  
std                      2.573082  
min                   2004.000000  
25%                   2018.000000  
50%                   2019.000000  
75%                   2020.000000  
max                   2023.000000

In [6]:
df.dtypes

Application_ID                object
Current City                  object
Python (out of 3)              int64
R Programming (out of 3)       int64
Data Science (out of 3)        int64
Other skills                  object
Institute                     object
Degree                        object
Stream                        object
Current Year Of Graduation     int64
Performance_PG                object
Performance_UG                object
Performance_12                object
Performance_10                object
dtype: object

In [7]:
df = df.astype({"Degree" : str}, {"Other skills" : str}) #To change the datatype of those two columns to string

we will create a new column 'skills' from the 'Other skills' column and display the matched skills we want for our data scientist role:, i.e:- 'Machine Learning', 'Deep Learning', 'nlp', 'ms-excel', 'sql', 'nosql', 'statistical data analysis','aws'


df['skills'] = (df['Other skills'].str.lower().str.extractall('(machine learning|deep learning|nlp|ms-excel|sql|nosql|statistical data analysis|aws)')
                      .count(level = 0)  # .size() to count duplicates
              )

In [8]:
def skills_counter(df):
    df['skills'] = (df['Other skills'].str.lower().str.extractall('(machine learning|deep learning|nlp|ms-excel|sql|nosql|statistical data analysis|aws)')
                      .groupby(level=0)
                      .nunique()  # .size() to count duplicates
              )


In [9]:
skills_counter(df)

In [10]:
df

Application_ID Current City  Python (out of 3)  R Programming (out of 3)  \
0           DS0001    Bangalore                  1                         0   
1           DS0002       Mumbai                  2                         1   
2           DS0003       Mumbai                  2                         0   
3           DS0004      Dhanbad                  2                         0   
4           DS0005    Bangalore                  2                         0   
..             ...          ...                ...                       ...   
606         DS0607         Pune                  2                         2   
607         DS0608    Bangalore                  1                         0   
608         DS0609        Delhi                  2                         2   
609         DS0610    Bangalore                  3                         2   
610         DS0611     Thrissur                  2                         2   

     Data Science (out of 3)  \
0                          3   
1                          2   
2                          0   
3                          2   
4                          0   
..                       ...   
606                        2   
607                        0   
608                        0   
609                        2   
610                        0   

                                          Other skills  \
0    Machine Learning, Arduino, C Programming, CSS,...   
1    AutoCAD, MS-Office, Machine Learning, Microsof...   
2    C++ Programming, Data Structures, Image Proces...   
3    Algorithms, C++ Programming, Data Structures, ...   
4                       MS-Word, Python, SQL, MS-Excel   
..                                                 ...   
606  AutoCAD, MS-Excel, Deep Learning, Machine Lear...   
607                        MS-Excel, HTML, Python, SQL   
608  Data Analytics, Machine Learning, Natural Lang...   
609  Data Analytics, Machine Learning, Python, C Pr...   
610  Computer Vision, Data Analytics, Image Process...   

                                             Institute  \
0                         Global Academy of Technology   
1    Aegis School Of Business, Data Science, Cyber ...   
2                                         VJTI, Mumbai   
3                                    IIT (ISM) Dhanbad   
4                                                 Vvce   
..                                                 ...   
606                                   Defour Analytics   
607  International Institute Of Information Technology   
608                                 Quantum University   
609        College Of Applied Science IHRD, Malampuzha   
610                                             Amrita   

                              Degree                                  Stream  \
0      Bachelor of Engineering (B.E)  Electrical and Electronics Engineering   
1                                nan                                     PGP   
2    Bachelor of Technology (B.Tech)                     Information Systems   
3                  Integrated M.Tech               Mathematics and Computing   
4      Bachelor of Engineering (B.E)           Electronics and Communication   
..                               ...                                     ...   
606    Bachelor of Engineering (B.E)                            Data Science   
607       PG Diploma in Data Science                            Data Science   
608  Bachelor of Technology (B.Tech)                  Mechanical Engineering   
609       Bachelor of Science (B.Sc)        Computer Science And Mathematics   
610    Master of Technology (M.Tech)                 E Learning Technologies   

     Current Year Of Graduation Performance_PG Performance_UG Performance_12  \
0                          2019            NaN        7.73/10            NaN   
1                          2020            NaN      68.00/100            NaN   
2                          2018            NaN        8.85/10  

In [11]:
degree = []

for words in df['Degree'].values:
    
    if "m.tech" in words.lower():
        degree.append("M.tech")
    elif "m.sc" in words.lower():
        degree.append("M.Sc")
    elif "b.tech" in words.lower():
        degree.append("B.Tech")
    elif "pg" in words.lower():
        degree.append("PG")
    elif "b.e" in words.lower():
        degree.append("B.Tech")
    else:
        degree.append(np.nan)
        

In [12]:
df.drop(columns = 'Degree', inplace = True)
Degree = pd.DataFrame(degree)
df.insert(7, "Degree", Degree)

In [13]:
df.rename(columns = {"Python (out of 3)": "Python",
                     "R Programming (out of 3)" : "R",
                     "Data Science (out of 3)" : "DS",
                     "Current Year Of Graduation" : "Year"},
                          inplace = True)

In [14]:
# Taking rows where Degree value is not na
df = df[df['Degree'].notna()]

# Taking rows where Year value is max 2020.
df = df.loc[df['Year'] <= 2020]

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 396 entries, 0 to 610
Data columns (total 15 columns):
Application_ID    396 non-null object
Current City      396 non-null object
Python            396 non-null int64
R                 396 non-null int64
DS                396 non-null int64
Other skills      389 non-null object
Institute         396 non-null object
Degree            396 non-null object
Stream            381 non-null object
Year              396 non-null int64
Performance_PG    64 non-null object
Performance_UG    354 non-null object
Performance_12    230 non-null object
Performance_10    220 non-null object
skills            350 non-null float64
dtypes: float64(1), int64(4), object(10)
memory usage: 49.5+ KB


### Insights
After applying the given condition that, Applicant should have a M.Sc / M.Tech / B.Tech/ B.E / PG degree , and graduation year should be max year 2020. We can see that out of 611 Applications only 383 Applications are shorlisted. So 62% selected for the further analysis based on their Degree and Year of Graduation.

In [16]:
def f(row):
    if row['Degree'] == 'B.Tech' and row['Year'] == 2020:
        val = 10
    elif row['Degree'] == 'B.Tech' and row['Year'] == 2019:
        val = 8
    elif row['Degree'] == 'B.Tech' and row['Year'] <= 2018:
        val = 5
    elif row['Degree'] == 'M.Tech' and row['Year'] == 2020:
        val = 7
    elif row['Degree'] == 'M.Tech' and row['Year'] <= 2019:
        val = 3
    elif row['Degree'] == 'M.Sc' and row['Year'] == 2020:
        val = 7
    elif row['Degree'] == 'M.Sc' and row['Year'] <= 2019:
        val = 3
    elif row['Degree'] == 'PG' and row['Year'] == 2020:
        val = 7
    elif row['Degree'] == 'PG' and row['Year'] <= 2019:
        val = 3
    else:
        val = 0
    return val

In [17]:
df['Degree_Year_Weightage'] = df.apply(f, axis = 1)

# Data Modification

As per the conditions:
__________________________________________________

If Python Value is 1 then it's weightage value is 3  
If Python Value is 2 then it's weightage value is 7  
If Python Value is 3 then it's weightage value is 10  
__________________________________________________

If R Value is 1 then it's weightage value is 3  
If R Value is 2 then it's weightage value is 7  
If R Value is 3 then it's weightage value is 10  
__________________________________________________

If DS Value is 1 then it's weightage value is 3  
If DS Value is 2 then it's weightage value is 7  
If DS Value is 3 then it's weightage value is 10  
__________________________________________________

Also Other Skills (i.e - Machine Learning, Deep Learning, NLP, Statistical Data Analysis, AWS, SQL, NoSQL, Excel) holds 3 weightage for each.

In [18]:
df['Python'].replace(3, 10, inplace = True)
df['Python'].replace(1, 3, inplace = True)
df['Python'].replace(2, 7, inplace = True)

df['R'].replace(3, 10, inplace = True)
df['R'].replace(1, 3, inplace = True)
df['R'].replace(2, 7, inplace = True)

df['DS'].replace(3, 10, inplace = True)
df['DS'].replace(1, 3, inplace = True)
df['DS'].replace(2, 7, inplace = True)

In [19]:
df['skills'].replace(np.nan, 0.0, inplace = True)
df = df.astype({"skills" : int})
df['skills'] = df['skills'].apply(lambda x: x * 3)

In [20]:
df.set_index(['Application_ID','Degree','Year'], inplace = True)

In [21]:
# creating a new column 'Total Score' and sum the scores of Python, R, Data Science(DS),Skills weightage,Degree_year_weightage.

df['Total Score'] = df.sum(axis = 1)

# Final Condition

If the Total Score is >= 40 then the Applicants profile has been shortlisted for Data Scientist, else Applicants profile did not qualify for further discussion.

In [22]:
def score(row):
    if row['Total Score'] >= 40:
        val = "Congratulations"
    else:
        val = "Sorry"
    return val

In [23]:
df['Results'] = df.apply(score, axis = 1)

# Chechking how many Applicants profile has been selected

In [24]:
df_shortlisted = df.loc[df['Results'] == "Congratulations"]

In [25]:
df_shortlisted

Current City  Python   R  DS  \
Application_ID Degree Year                                
DS0077         B.Tech 2019    Bangalore      10   7  10   
DS0136         B.Tech 2019      Chennai       7   7   7   
DS0163         B.Tech 2019    Bangalore       7   7  10   
DS0172         B.Tech 2020         Pune       7   7   7   
DS0180         B.Tech 2019    Bangalore      10   3  10   
DS0187         B.Tech 2020    Bangalore      10   7   7   
DS0273         B.Tech 2020        Delhi       7  10   7   
DS0278         B.Tech 2019    Bangalore      10  10  10   
DS0294         PG     2019    Hyderabad      10  10  10   
DS0308         B.Tech 2020    Bangalore       7   7   7   
DS0320         B.Tech 2018    Bangalore      10  10  10   
DS0381         B.Tech 2019    Bangalore      10   7   7   
DS0421         B.Tech 2019      Chennai       7   7   7   
DS0426         B.Tech 2019    Bangalore      10  10  10   
DS0435         B.Tech 2019       Nagpur       7   7   3   
DS0447         B.Tech 2020    Raebareli       7   7   7   
DS0468         B.Tech 2019    Bangalore      10  10  10   
DS0516         B.Tech 2019    Bangalore      10   7  10   
DS0567         B.Tech 2019      Kolkata       7   7  10   
DS0574         B.Tech 2019    Bangalore       7   7   7   
DS0595         B.Tech 2020    Bangalore      10   7   3   
DS0607         B.Tech 2020         Pune       7   7   7   

                                                                 Other skills  \
Application_ID Degree Year                                                      
DS0077         B.Tech 2019  Data Science, Deep Learning, Machine Learning,...   
DS0136         B.Tech 2019  Artifical Intelligence, Computer Vision, Deep ...   
DS0163         B.Tech 2019  Data Science, Linear Programming, MS-Excel, Ma...   
DS0172         B.Tech 2020  AutoCAD, MS-Excel, Deep Learning, Machine Lear...   
DS0180         B.Tech 2019  Algorithms, C Programming, C++ Programming, Ma...   
DS0187         B.Tech 2020  MS-Excel, MS-Office, Python, C Programming, C+...   
DS0273         B.Tech 2020  R Programming, Computer Vision, Data Analytics...   
DS0278         B.Tech 2019  Adobe Photoshop, CSS, Computer Vision, Data An...   
DS0294         PG     2019  Artifical Intelligence, MS-Excel, Machine Lear...   
DS0308         B.Tech 2020  C++ Programming, MS-Excel, MySQL, HTML, JSON, ...   
DS0320         B.Tech 2018  Deep Learning, Machine Learning, Python, R Pro...   
DS0381         B.Tech 2019  Data Analytics, Data Science, English Proficie...   
DS0421         B.Tech 2019  Artifical Intelligence, Computer Vision, Deep ...   
DS0426         B.Tech 2019  Artifical Intelligence, Data Analytics, Deep L...   
DS0435         B.Tech 2019  Data Analytics, Data Science, Machine Learning...   
DS0447         B.Tech 2020  C Programming, Data Structures, Deep Learning,...   
DS0468         B.Tech 2019  Adobe Photoshop, CSS, Computer Vision, Data An...   
DS0516         B.Tech 2019  Data Science, Deep Learning, Machine Learning,...   
DS0567         B.Tech 2019  Deep Learning, Machine Learning, OpenCV, C++ P...   
DS0574         B.Tech 2019  JavaScript, SQL, C Programming, CSS, Data Anal...   
DS0595         B.Tech 2020  MS-Excel, MS-Office, Python, C Programming, C+...   
DS0607         B.Tech 2020  AutoCAD, MS-Excel, Deep Learning, Machine Lear...   

                                                                    Institute  \
Application_ID Degree Year                                                      
DS0077         B.Tech 2019                          IES College of Technology   
DS0136         B.Tech 2019                   A B R Engineering And Technolagy   
DS0163         B.Tech 2019                   Gujarat Technological University   
DS0172         B.Tech 2020                                   Defour Analytics   
DS0180         B.Tech 2019                St. Joseph's College of Engineering   
DS0187         B.Tech 2020  Centurion University of Technology and Management   
DS0273         B.

In [26]:
df_shortlisted.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 22 entries, (DS0077, B.Tech, 2019) to (DS0607, B.Tech, 2020)
Data columns (total 15 columns):
Current City             22 non-null object
Python                   22 non-null int64
R                        22 non-null int64
DS                       22 non-null int64
Other skills             22 non-null object
Institute                22 non-null object
Stream                   22 non-null object
Performance_PG           0 non-null object
Performance_UG           20 non-null object
Performance_12           12 non-null object
Performance_10           13 non-null object
skills                   22 non-null int64
Degree_Year_Weightage    22 non-null int64
Total Score              22 non-null int64
Results                  22 non-null object
dtypes: int64(6), object(9)
memory usage: 6.1+ KB
